File: card-lang/src/calculator/evaluator/parts/functions/trig.rs
use std::collections::HashMap;

use crate::types::Evaluator::Funk;
use crate::types::Evaluator::FunkRunner;

use crate::types::ReusedStructs::Node;
use crate::types::ReusedStructs::NodeType;

pub struct SinRunner{}
impl FunkRunner for SinRunner {
  fn run(self, args: Vec<Node> )-> Node {
    let nums = extractNumbers(args);
    return packResult(nums[0].sin());
  }
}

pub struct CosRunner{}
impl FunkRunner for CosRunner {
  fn run(self, args: Vec<Node> )-> Node {
    let nums = extractNumbers(args);
    return packResult(nums[0].cos());
  }
}

pub struct TanRunner{}
impl FunkRunner for TanRunner {
  fn run(self, args: Vec<Node> )-> Node {
    let nums = extractNumbers(args);
    return packResult(nums[0].tan());
  }
}

pub struct ASinRunner{}
impl FunkRunner for ASinRunner {
  fn run(self, args: Vec<Node> )-> Node {
    let nums = extractNumbers(args);
    return packResult(nums[0].asin());
  }
}

pub struct ACosRunner{}
impl FunkRunner for ACosRunner {
  fn run(self, args: Vec<Node> )-> Node {
    let nums = extractNumbers(args);
    return packResult(nums[0].acos());
  }
}

pub struct ATanRunner{}
impl FunkRunner for ATanRunner {
  fn run(self, args: Vec<Node> )-> Node {
    let nums = extractNumbers(args);
    return packResult(nums[0].atan());
  }
}



pub fn extractNumbers(nums: Vec<Node>) -> Vec<f64> {
  return nums.iter().map(|node|-> f64 {
    if !matches!(node.node_type, NodeType::ValueNode) {
      panic!("an argument node_type is not ValueNode");
    } 
    node.values.unwrap()["value"].parse::<f64>().unwrap()
  }).collect();
}

pub fn packResult(num: f64)->Node{
  Node {
    node_type: NodeType::ValueNode,
    values: Some(HashMap::from([
      ("value".to_string(), num.to_string())
    ])),
    branches: None,
    args: None
  }
}

File: card-lang-typescript/src/calculator/evaluator/parts/functions/trig.ts
import { FunkRunner } from "../../../../types/Evaluator";
import { LangNode, ValueNode } from "../../../../types/ReusedStructs";
import { extractArray, packResult } from "../../utils";

export class SinRunner implements FunkRunner {
  run(args: Array<LangNode> ): ValueNode {
    let nums = extractArray(args);
    return packResult(Math.sin(nums[0]));
  }
}

export class CosRunner implements FunkRunner {
  run(args: Array<LangNode> ): ValueNode {
    let nums = extractArray(args);
    return packResult(Math.cos(nums[0]));
  }
}

export class TanRunner implements FunkRunner {
  run(args: Array<LangNode> ): ValueNode {
    let nums = extractArray(args);
    return packResult(Math.tan(nums[0]));
  }
}

export class ASinRunner implements FunkRunner {
  run(args: Array<LangNode> ): ValueNode {
    let nums = extractArray(args);
    return packResult(Math.asin(nums[0]));
  }
}

export class ACosRunner implements FunkRunner {
  run(args: Array<LangNode> ): ValueNode {
    let nums = extractArray(args);
    return packResult(Math.acos(nums[0]));
  }
}

export class ATanRunner implements FunkRunner {
  run(args: Array<LangNode> ): ValueNode {
    let nums = extractArray(args);
    return packResult(Math.atan(nums[0]));
  }
}
